### PR TITLE
fix(lib/genesis): Update missing and incorrect fields in genesis file.

### LINKED ISSUE #1580: determine and update potentially missing storage genesis entries
## Task summary

- currently we only set a few storage entries at genesis using build-spec: babe authorities, grandpa authorities, balances, and `:code`
- however there are likely many more storage entries that need to be set at genesis that are currently unknown
- for example: https://github.com/ChainSafe/gossamer/pull/1573 adding a storage field fixes various bugs, however it is unknown what the storage key is and how the value is generated
- need to figure out what the key/value represents and how to create it and add it to `build-spec`
- same for other potentially missing keys

### PR DESCRIPTION
**Pull request by arijitAD**
## Changes

- Update missing and incorrect fields in the genesis file.

## Tests

```

```

## Issues

- closes #1580

### CODE REVIEW

**Comment by edwardmack** [comment]
How did you determine what fields are missing are incorrect?

**Comment by arijitAD** [comment]
> How did you determine what fields are missing are incorrect?

I read through the substrate code and generated the substrate genesis file and compared its values with ours.

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: lib/genesis/helpers.go [hunk 1]
```diff
@@ -37,6 +37,8 @@ import (
 	"github.com/ChainSafe/gossamer/pkg/scale"
 )
 
+type accountAddr [32]byte
```
[the remark this answers is not in the document]
Comment (noot) [reply lib/genesis/helpers.go:40]: `common.Hash` represents a hash so it's good to have a separate type for something that doesn't represent a hash

File: lib/genesis/helpers.go [hunk 2]
```diff
@@ -157,23 +159,126 @@ type keyValue struct {
...
+	case "PhragmenElection":
+		pe := &phragmenElection{}
+		if err = json.Unmarshal(jsonBody, pe); err != nil {
+			return false, err
+		}
+
+		err = generateKeyValue(pe, k, res)
+		if err != nil {
+			return false, err
+		}
```
Comment (noot) [inline lib/genesis/helpers.go:248]: this code seems to be the same for every case, is it possible to move it out of the switch case?

File: lib/genesis/helpers.go [hunk 3]
```diff
@@ -232,33 +336,249 @@ func buildRawArrayInterface(a []interface{}, kv *keyValue) error {
...
+		for _, add := range t {
+			accID := crypto.PublicAddressToByteArray(common.Address(add))
+			encode = append(encode, accID...)
+		}
+		encode, err = scale.Marshal(encode)
+		if err != nil {
+			return nil, err
+		}
+	case string:
+		if t == "NotForcing" { //TODO: check if this is ok
```
[the remark this answers is not in the document]
Comment (noot) [reply lib/genesis/helpers.go:396]: can you document what this does?

File: lib/genesis/helpers.go [hunk 4]
```diff
@@ -232,33 +336,249 @@ func buildRawArrayInterface(a []interface{}, kv *keyValue) error {
...
+}
+
+func generateContractKeyValue(c *contracts, prefixKey string, res map[string]string) error {
+	var (
+		key string
+		err error
+	)
+	// First field of contract is the storage key
+	val := reflect.ValueOf(c)
+	if k := reflect.Indirect(val).Type().Field(0).Name; k == "CurrentSchedule" {
```
[the remark this answers is not in the document]
Comment (noot) [reply lib/genesis/helpers.go:438]: maybe use `strings.Equal`?

File: lib/genesis/helpers.go [hunk 5]
```diff
@@ -232,33 +336,249 @@ func buildRawArrayInterface(a []interface{}, kv *keyValue) error {
...
+}
+
 func formatKey(kv *keyValue) (string, error) {
 	switch {
-	case reflect.DeepEqual([]string{"grandpa", "authorities"}, kv.key):
+	case reflect.DeepEqual([]string{"Grandpa", "authorities"}, kv.key):
 		kb := []byte(`:grandpa_authorities`)
 		return common.BytesToHex(kb), nil
-	case reflect.DeepEqual([]string{"system", "code"}, kv.key):
+	case reflect.DeepEqual([]string{"System", "changesTrieConfig", "code"}, kv.key):
```
Comment (noot) [inline lib/genesis/helpers.go:480]: I'm a bit confused on the addition of `changesTrieConfig` here, is it basically just another potential field under `System`?

File: lib/genesis/helpers.go [hunk 6]
```diff
@@ -232,33 +336,249 @@ func buildRawArrayInterface(a []interface{}, kv *keyValue) error {
...
+				res[common.BytesToHex(append(append(moduleName, nextKeyHash...), append(accIDHash, acc...)...))] = common.BytesToHex(acc)
+			case map[string]interface{}:
+				var storagePrefixKey []byte
+				storagePrefixKey, err = common.Twox128Hash([]byte("KeyOwner"))
+				if err != nil {
+					return err
+				}
+
+				storagePrefixKey = append(moduleName, storagePrefixKey...)
+				validatorAccID := crypto.PublicAddressToByteArray(common.Address("5GNJqTPyNqANBkUVMN1LPPrxXnFouWXoe2wNSmmEoLctxiZY"))
```
Comment (noot) [inline lib/genesis/helpers.go:530]: who's address is this / where is it from?

File: lib/genesis/helpers.go [hunk 7]
```diff
@@ -232,33 +336,249 @@ func buildRawArrayInterface(a []interface{}, kv *keyValue) error {
...
+				}
+			}
 		}
-		return common.BytesToHex(append(prefix, keydata...)), nil
 	}
+
+	return nil
+}
+
+func generateAddHash(accAddr, key string) ([]byte, error) {
```
Comment (noot) [inline lib/genesis/helpers.go:567]:
```suggestion
func generateAddressHash(accAddr, key string) ([]byte, error) {
```

File: lib/genesis/pallet.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
+package genesis
+
+type contracts struct {
+	CurrentSchedule struct {
```
Comment (noot) [inline lib/genesis/pallet.go:4]: it's a private struct so should the fields inside also be private (or are they needed for encoding/decoding and such)?

**Review by noot — COMMENTED** [review]
overall looks good! would be nice to get more unit test cases in here for the changes

File: lib/genesis/helpers.go [hunk 9]
```diff
@@ -157,23 +159,126 @@ type keyValue struct {
...
 }
 
+func generatePalletKeyValue(k string, v map[string]interface{}, res map[string]string) (bool, error) {
+	jsonBody, err := json.Marshal(v)
+	if err != nil {
+		return false, err
+	}
+
+	switch k {
+	case "Society":
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply lib/genesis/helpers.go:169]: Done

File: lib/genesis/helpers.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (arijitAD) [reply lib/genesis/helpers.go]: `strings` package doesn't have `Equal()`

File: lib/genesis/helpers.go [hunk 10]
```diff
@@ -232,33 +336,249 @@ func buildRawArrayInterface(a []interface{}, kv *keyValue) error {
...
+	return nil
+}
+
+func generateKeyValue(s interface{}, prefixKey string, res map[string]string) error {
+	val := reflect.ValueOf(s)
+	n := reflect.Indirect(val).NumField()
+	for i := 0; i < n; i++ {
+		val := reflect.ValueOf(s)
+		storageKey := reflect.Indirect(val).Type().Field(i).Name
+		if storageKey == "Phantom" { //TODO: figure out what to do with Phantom as its value is null
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply lib/genesis/helpers.go]: Done

File: lib/genesis/helpers.go (hunk 2 above)
Comment (arijitAD) [reply lib/genesis/helpers.go]: Done

File: lib/genesis/helpers.go (hunk 5 above)
Comment (arijitAD) [reply lib/genesis/helpers.go:457]: Currently, we are not using `changesTrieConfig `. I have removed it and added a todo.

File: lib/genesis/helpers.go (hunk 6 above)
Comment (arijitAD) [reply lib/genesis/helpers.go]: Removed. I hard-coded it for testing.

File: lib/genesis/helpers.go (hunk 7 above)
Comment (arijitAD) [reply lib/genesis/helpers.go]: Done

File: lib/genesis/pallet.go (hunk 8 above)
Comment (arijitAD) [reply lib/genesis/pallet.go:4]: The struct needs to be public for encoding/decoding

File: lib/genesis/helpers.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (arijitAD) [reply lib/genesis/helpers.go]: Removed the todo and added the code.

File: lib/genesis/helpers.go [hunk 11]
```diff
@@ -232,33 +316,257 @@ func buildRawArrayInterface(a []interface{}, kv *keyValue) error {
...
+				}
+
+				validatorAccID = crypto.PublicAddressToByteArray(common.Address(t))
+				var accIDHash []byte
+				accIDHash, err = common.Twox64(validatorAccID)
+				if err != nil {
+					return err
+				}
+
+				res[common.BytesToHex(append(append(moduleName, nextKeyHash...), append(accIDHash, validatorAccID...)...))] = common.BytesToHex(validatorAccID)
```
Comment (EclesioMeloJunior) [inline lib/genesis/helpers.go:511]: spliting those appends will be more readable IMO 
```go
prefix := bytes.Join([][]byte{moduleName, nextKeyHash}, []byte{})
sufix := bytes.Join([][]byte{accIDHash, validatorAccID}, []byte{})

res[common.BytesToHex(prefix, sufix...)] = common.BytesToHex(validatorAccID)
```

File: lib/genesis/helpers.go [hunk 12]
```diff
@@ -232,33 +316,257 @@ func buildRawArrayInterface(a []interface{}, kv *keyValue) error {
...
+	case uint64:
+		encode, err = scale.Marshal(t)
+		if err != nil {
+			return nil, err
+		}
+	case uint32:
+		encode, err = scale.Marshal(t)
+		if err != nil {
+			return nil, err
+		}
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply lib/genesis/helpers.go:344]: Done

File: lib/genesis/helpers.go (hunk 11 above)
Comment (arijitAD) [reply lib/genesis/helpers.go]: Done

**Review by EclesioMeloJunior — APPROVED** [review]

**Pull request merged**